.changeset/poor-goats-promise.md

```diff
@@ -0,0 +1,19 @@
+---
+"@cloudflare/pages-shared": patch
+"wrangler": patch
+---
+
+fix: Allow unsetting an header before setting again in the same rule in Pages `_headers`
+
+Previously, it wasn't possible to unset a header with the `! Header-Name` operator, before setting a new value. You could only do this in separate rules. Now you can unset any existing value and set a new one, all in the same rule.
+
+For example,
+
+```txt
+/*
+	Header-Name: Old value
+
+/foo
+	! Header-Name
+	Header-Name: New value
```

Comment:
Is this syntax a standard?
If not, I wonder if we could invent something the means "replace all these headers with this one?"
Perhaps something like:

```
/foo
  * Header-Name: New Value
```

where the `*` means delete any previous headers with this name before creating this one?

This is obviously less verbose, but also, IMO, more intuitive.

Reply:
https://developers.cloudflare.com/pages/platform/headers#detach-a-header

There is no standard for `_headers` as of yet (but hopefully one day!). I consider your suggestion a feature request for

```txt
/foo
  * Header-Name: New Value
```

to be a shorthand for

```txt
/foo
  ! Header-Name
  Header-Name: New Value
```

But I do still think the second probably makes sense all on its own. We do support unsetting, but the behavior today when you set and unset in the same rule is that the unsetting runs last, thereby making the setting a noop. This seems "buggy" to me, and we had one report of confusion around this.

Reply:
I also feel slightly awkward that `! Header-Name` unsets "all" headers with that name. Whereas `Header-Name` only adds headers. I wish there was some way to distinguish.

Reply:
We _kinda_ have similar behavior with that today too.

Forgetting unsetting for a moment, if a header is set in the `_headers`, it will completely override any existing values for that header name (i.e. any Pages-provided header (e.g. `Cache-Control`) is dropped if you specify it in `_headers`). This kinda makes sense because someone is seizing control of this header and wants to override anything that was there before.

But, if there are multiple instances of that header in the `_headers` file, they concatenate. This is useful to build up values as you nest your pages. For example:

```
/*
  Link: </my-site-styles.css>

/blog/*
  Link: </some-blog-specific-styles.css>
```

This will produce `Link:  </my-site-styles.css>, </some-blog-specific-styles.css>` for requests to `/blog/foo`.

So to me, a blunt `! unset-me-completely` is probably pretty okay if you're needing to, say, override something because you have a completely different type of sub-app. For example:

```
/*
  Link: </my-site-styles.css>

/admin/*
  ! Link
  Link: </admin-only-styles.css>
```

where `/admin` is maybe backed by a different framework.

Either way, we can't break the behavior of `! unset` now, so this PR is exclusively about fixing the "bug" that we see when trying to unset and set a new value all in the same rule.

Reply:
Hey what about this:

```
/*
  Header-1: value-one
  Header-2: value-two

/admin/*
  ! Header-1
  ! Header-2: value-three
```

This would result in Header-1 being removed completely and Header-2 being replaced with the new value.